### PR TITLE
Change how "additional_special_tokens" argument in the ".from_pretrained" method of the tokenizer is taken into account

### LINKED ISSUE #12533: The "additional_special_tokens" argument in the ".from_pretrained" method of the tokenizer is not necessarily taken into account.
## Environment info

- `transformers` version: 4.8.2
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.10
- PyTorch version (GPU?): 1.9.0+cu102 (False)
- Tensorflow version (GPU?): 2.5.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help

@LysandreJik 
@sgugger 

## Information

Model I am using (Bert, XLNet ...): Tapas

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

To easily reproduce the behavior, you can directly open this [Google colab](https://colab.research.google.com/drive/1zKvkdmH2blYO4dx1s6OojTkDQcBGd35L?usp=sharing)

1. With the latest version of transformers, initialized a Tapas tokenizer with a new added token `"<special>"`
```
from transformers import TapasTokenizer

added_tokens = ["<special>"]

pretrained_name = "google/tapas-base"
tokenizer_p = TapasTokenizer.from_pretrained(
    pretrained_name, additional_special_tokens=added_tokens
)
```
2. Check if the new added token has been taken into account
```
"<special>" in tokenizer_p.get_vocab()
```
Output: `False`
```
tokenizer_p.additional_special_tokens # I would expect "<special>" in the list
```
Output: `['[EMPTY]']`
```
special_token_id = tokenizer_p.convert_tokens_to_ids(["<special>"])
print(tokenizer_p.convert_ids_to_tokens(special_token_id))
```
Output: `['[UNK]']`
3. See how an input containing `"<special>"` is tokenized
```
import pandas as pd 
query = "Hey this is a <special> token"
data = [
    ["Pos"],
    ["1"],
    ["2"],
]
table = pd.DataFrame.from_records(data[1:], columns=data[0])
p_output = tokenizer_p.encode(table, query)
```
Output: `['[CLS]', 'hey', 'this', 'is', 'a', '<', 'special', '>', 'token', '[SEP]', 'po', '##s', '1', '2']`

## Expected behavior

I would have hoped that the special token would be taken into account. In other words, I would have hoped to have the following outputs for each of the commands:
```
"<special>" in tokenizer_p.get_vocab()
```
Output: `True`
```
tokenizer_p.additional_special_tokens # I would expect "<special>" in the list
```
Output: `['<special>', '[EMPTY]']`
```
special_token_id = tokenizer_p.convert_tokens_to_ids(["<special>"])
print(tokenizer_p.convert_ids_to_tokens(special_token_id))
```
Output: `['<special>']`
```
import pandas as pd 
query = "Hey this is a <special> token"
data = [
    ["Pos"],
    ["1"],
    ["2"],
]
table = pd.DataFrame.from_records(data[1:], columns=data[0])
p_output = tokenizer_p.encode(table, query)
```
Output: `['[CLS]', 'hey', 'this', 'is', 'a', '<special>', 'token', '[SEP]', 'po', '##s', '1', '2']`

## Explanation ideas

I have the impression that when the `"additional_special_tokens"` key is filled in the `special_tokens_map.json` file then only the associated list is kept for the `additional_special_tokens` attribute. The content in the `additional_special_tokens` argument in ` TapasTokenizer.from_pretrained(pretrained_name, additional_special_tokens=added_tokens)` is then ignored. 

Is this the behavior we want to have? 
From the test [here](https://github.com/huggingface/transformers/blob/master/tests/test_tokenization_common.py#L3125) I have the impression that it is not -- it would fail on "google/tapas-base" of TapasTokenizer which is not tested at the moment but will be as part of the issue #12535.

### PR DESCRIPTION
**Pull request by SaulLu**
# What does this PR do?

This PR is a proposal for the issue #12533.

## Motivation

This change in behavior is motivated in particular by the [`test_special_tokens_initialization` test in `test_tokenization_common`](https://github.com/huggingface/transformers/blob/v4.9.2/tests/test_tokenization_common.py#L3132). 

This test consists in loading thanks to the method `from_pretrained` a tokenizer (on the hub)  and adding in the arguments `additional_special_tokens=[AddedToken("<special>", lstrip=True)]` then to check that this new special token `"<special>"` was well added.

```python
added_tokens = [AddedToken("<special>", lstrip=True)]
tokenizer_r = self.rust_tokenizer_class.from_pretrained(
                    pretrained_name, additional_special_tokens=added_tokens, **kwargs
r_output = tokenizer_r.encode("Hey this is a <special> token")
special_token_id = tokenizer_r.encode("<special>", add_special_tokens=False)[0]
self.assertTrue(special_token_id in r_output)
```

This test does not currently work in one case: if the repository on the hub contains a `"special_tokens_map.json"` file that defines the value of `"additional_special_tokens"` then the new argument `additional_special_tokens=[AddedToken("<special>", lstrip=True)]` added in the `.from_pretrained` method is ignored.

## New behavior introduce by this PR

This PR introduces a change that gives priority to arguments provided in the `.from_pretrained` method. Understand that if there is a competition between the definition of an argument in the `.from_pretrained` method and in the repository files, then the definition of the argument in the method will be chosen.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. In particular:
@LysandreJik and @sgugger - It would be awesome if you could validate the new behavior and give your opinion on the PR
@NielsRogge - It would be great if you could look at the proposed change in the tokenization of Luke and possibly the test of Canine
@patil-suraj - It would be great if you could look at the proposed change in the tokenization of M2M-100 and MBart50. For these 2 models, I have proposed changes so that the user can still define additional special tokens even if these templates already define additional special tokens corresponding to the language codes. This behavior was already present in `TokenizerMBart` but not in the tokenizers of these last models. If these are not desirable behaviors for these models, I can reverse the changes and just disable the corresponding tests for these models
@patrickvonplaten - it would be great if you could take a look at the change in the Wave2Vec2 processor test (if you agree with this change, there is the same function used in `test_tokenization_wave2vec2.py` that I didn't change)

Thank you for your time

## Failing Test
One test is currently failing `test_run_seq2seq_no_dist` in `test_trainer_ext.py`. At the moment I'm not sure that I see the link with the changes I introduced, but I'm still investigating.

EDIT: I think this test is failing do to a change in [sacrebleu](https://github.com/mjpost/sacrebleu/) which affects :hugs: datasets library. I understand that the problem is in process to be solved on the datasets side ([PR](https://github.com/huggingface/datasets/pull/2778))

### CODE REVIEW

File: src/transformers/models/luke/tokenization_luke.py [hunk 1]
```diff
@@ -210,8 +210,8 @@ def __init__(
             if isinstance(entity_token_2, str)
             else entity_token_2
         )
-        kwargs["additional_special_tokens"] = [entity_token_1, entity_token_2]
-        kwargs["additional_special_tokens"] += kwargs.get("additional_special_tokens", [])
+        kwargs["additional_special_tokens"] = kwargs.get("additional_special_tokens", [])
+        kwargs["additional_special_tokens"] += [entity_token_1, entity_token_2]
```
Comment (SaulLu) [inline src/transformers/models/luke/tokenization_luke.py:214]: Pinging to check I'm not missing something here

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -1862,6 +1862,12 @@ def convert_added_tokens(obj: Union[AddedToken, Any]):
             with open(special_tokens_map_file, encoding="utf-8") as special_tokens_map_handle:
                 special_tokens_map = json.load(special_tokens_map_handle)
             for key, value in special_tokens_map.items():
+                if key in kwargs and kwargs[key]:
+                    # This value has already been redefined by the kwargs
+                    # We keep this new value and ignore the one stored in the special_tokens_map_file
+
+                    continue
```
Comment (SaulLu) [inline src/transformers/tokenization_utils_base.py:1869]: I just want to draw attention to this part of the code which is the main change and which inverts the priority rules.

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! This is definitely cleaner for the multilingual tokenizers. The change in the rule makes complete sense and the new test looks great.

**Review by LysandreJik — APPROVED** [review]
This looks great. Thank you for working on it @SaulLu!

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me!

File: tests/test_tokenization_common.py [hunk 3]
```diff
@@ -3161,6 +3162,64 @@ def test_special_tokens_initialization(self):
...
+                # the following checks allow us to verify that our test works as expected, i.e. that the tokenizer takes
+                # into account the new value of additional_special_tokens given in the "tokenizer_config.json" and
+                # "special_tokens_map.json" files
+                tokenizer_without_change_in_init = tokenizer_class.from_pretrained(
+                    tmp_dir,
+                )
+                self.assertIn(
+                    "an_additional_special_token", tokenizer_without_change_in_init.additional_special_tokens
+                )
+                # self.assertIn("an_additional_special_token",tokenizer_without_change_in_init.get_vocab()) # ByT5Tokenization no vocab
```
Comment (NielsRogge) [inline tests/test_tokenization_common.py:3200]: Is this a line that still needs to be removed?

**Comment by NielsRogge** [comment]
LGTM! Thank you for making this cleaner.

File: tests/test_tokenization_common.py (hunk 3 above)
Comment (SaulLu) [reply tests/test_tokenization_common.py]: Good point! This line should not be commented as this test is overridden in `tests/test_tokenization_byt5.py`. I just changed it.

**Comment by LysandreJik** [comment]
Feel free to merge after solving the code quality issues!

**Pull request merged**